src/combinatorics/simple_graph/connectivity.lean

```diff
@@ -284,6 +334,48 @@ end
 @[simp] lemma length_eq_zero_iff {u : V} {p : G.walk u u} : p.length = 0 ↔ p = nil :=
 by cases p; simp
 
+section concat_rec
+
+variables
+  {motive : Π (u v : V), G.walk u v → Sort*}
+  (Hnil : Π {u : V}, motive u u nil)
+  (Hconcat : Π {u v w : V} (p : G.walk u v) (h : G.adj v w), motive u v p → motive u w (p.concat h))
+
+/-- Auxiliary definition for `simple_graph.walk.concat_rec` -/
+def concat_rec_aux : Π {u v : V} (p : G.walk u v), motive v u p.reverse
+| _ _ nil := Hnil
+| _ _ (cons h p) := eq.rec (Hconcat p.reverse (G.symm h) (concat_rec_aux p)) (reverse_cons h p).symm
+
+lemma concat_rec_aux_concat {u v w : V} (p : G.walk u v) (h : G.adj v w) :
+  concat_rec_aux @Hnil @Hconcat (p.concat h).reverse
+  == Hconcat p.reverse.reverse h (concat_rec_aux @Hnil @Hconcat p.reverse) :=
+begin
+  transitivity concat_rec_aux @Hnil @Hconcat (cons h.symm p.reverse),
+  { congr, simp },
+  simp [concat_rec_aux, rec_heq_iff_heq],
+end
+
+/-- Recursor on walks by inducting on `simple_graph.walk.concat`. -/
+@[elab_as_eliminator]
+def concat_rec {u v : V} (p : G.walk u v) : motive u v p :=
+eq.rec (concat_rec_aux @Hnil @Hconcat p.reverse) (reverse_reverse p)
```

Comment:
I could mimic [list.reverse_rec_on](https://github.com/leanprover-community/mathlib/blob/27ac3e1ad77a9f99379e3078eed4c78b3f9e5cf9/src/data/list/basic.lean#L944) to produce the following construction:
```suggestion
include Hnil Hconcat
/-- Recursor on walks by inducting on `simple_graph.walk.concat`. -/
@[elab_as_eliminator]
def concat_rec {u v : V} (p : G.walk u v) : motive u v p :=
begin
  rw ← reverse_reverse p,
  induction reverse p,
  { exact Hnil },
  case cons : _ _ _ _ q ih { rw reverse_cons, exact Hconcat _ _ (ih q.reverse) }
end
```
Is the current approach superior? The heterogeneous equality makes me dizzy ...

Reply:
Unless I'm mistaken, I believe that's precisely the same definition, just with tactics rather than `eq.rec`. I just wanted to make sure I had exact control of the terms for the proofs of the evaluation rules `concat_rec_nil` and `concat_rec_concat`. Too many recursors in mathlib do not have these.

Regarding `heq`, the proofs in this file finally made me understand why it's so nice. So long as you can keep proofs of type equality in your context, it lets you do fairly straightforward equational reasoning that lets you erase `eq.rec`/`cast`. My usual strategy has been to avoid it and to manipulate the `eq.rec`s manually.

Reply:
Yes, but my construction doesn't use the auxiliary lemma `concat_rec_aux` which depends on one more lemma. You could probably make it term-mode as well. Isn't that better?

Reply:
I folded in that `concat_rec_aux` lemma so now you don't have to see it, but I've kept this auxiliary definition.

There's nothing wrong with having auxiliary definitions. While I could fold this into `concat_rec` using an explicit recursor, I believe that would hurt readability.

(When it comes to these engineering questions, there's so single axis for what's "better" -- for me, I see that the current code works and it has a clean API, so spending any more time futzing with it would be time wasted.)